server/db/driver/pg/upgrades.go

```diff
@@ -0,0 +1,302 @@
+					startRate = midRate
+				} else {
+					startRate = endRate // from previous epoch with matches
+				}
+				endRate = midRate
+
+				// No book buy / sell depth (see bookVolumes in server/matcher).
+				_, err = reportStmtPrep.Exec(epochEnd, dur, matchVolume, quoteVolume,
+					highRate, lowRate, startRate, endRate)
+				if err != nil {
+					return err
+				}
+			}
+		} // range durs
+		log.Debugf("Processed %d matches doing %s in %s volume (%s in %s volume)", totalMatches,
+			strconv.FormatFloat(float64(totalVolume)/1e8, 'f', -1, 64), strings.ToUpper(dex.BipIDSymbol(mkt.Base)),
+			strconv.FormatFloat(float64(totalQVolume)/1e8, 'f', -1, 64), strings.ToUpper(dex.BipIDSymbol(mkt.Quote)))
+		return nil
+	}
+
+	for _, mkt := range mkts {
+		err = doMarketMatches(mkt)
+		if err != nil {
+			_ = tx.Rollback()
+			return err
+		}
+	}
+	return tx.Commit()
+}
+
+// DBVersion retrieves the database version from the meta table.
+func DBVersion(db *sql.DB) (ver uint32, err error) {
+	err = db.QueryRow(internal.SelectDBVersion).Scan(&ver)
+	return
+}
+
+func setDBVersion(db *sql.DB, ver uint32) error {
+	res, err := db.Exec(internal.SetDBVersion, ver)
+	if err != nil {
+		return err
+	}
+
+	n, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if n != 1 {
+		return fmt.Errorf("set the DB version in %d rows instead of 1", n)
+	}
+	return nil
+}
+
+func columnExists(db *sql.DB, schema, table, col string) (bool, error) {
+	var found bool
+	err := db.QueryRow(`SELECT EXISTS (
+			SELECT column_name
+			FROM information_schema.columns
+			WHERE table_schema = $1 AND table_name = $2 AND column_name = $3
+		);`, schema, table, col).Scan(&found)
+	if err != nil {
+		return false, err
+	}
+	return found, nil
+}
+
+func upgradeDB(db *sql.DB) error {
+	// Get the DB version from the meta table. Nonexistent schema_version column
+	// implies v0, the upgrade from which adds the schema_version column.
+	var current uint32
+	found, err := columnExists(db, "public", metaTableName, "schema_version")
+	if err != nil {
+		return err
+	}
+	if found {
+		current, err = DBVersion(db)
+		if err != nil {
+			return fmt.Errorf("failed to get DB version: %w", err)
+		}
+	} // else upgrade to v1 creates schema_version
+
+	if current == dbVersion {
+		log.Infof("DCRDEX database ready at version %d", dbVersion)
+		return nil // all upgraded
+	}
+
+	if current > dbVersion {
+		return fmt.Errorf("current DB version %d is newer than highest recognized version %d",
+			current, dbVersion)
+	}
+
+	log.Infof("Upgrading DB scheme from %d to %d", current, len(upgrades))
+	for i, up := range upgrades[current:] {
+		targetVer := current + uint32(i) + 1
+		log.Debugf("Upgrading DB scheme to %d...", targetVer)
+		if err = up(db); err != nil {
+			return fmt.Errorf("failed to upgrade to db version %d: %w", targetVer, err)
+		}
+
+		if err = setDBVersion(db, targetVer); err != nil {
+			return fmt.Errorf("failed to set new DB version %d: %w", targetVer, err)
+		}
+	}
```

Comment:
If setting the db version fails, we'll have an upgraded database with an older version number. Can avoid this by using a transaction for each upgrade, pass the transaction to each upgrade function, after the upgrade function returns without error, set the db version before committing. Any error should rollback all changes.

Reply:
Good call!  Hopefully no upgrades push what can be done in a single transaction.

Reply:
I didn't encode it all in a transaction yet, but otherwise ready for review.  I'm actually thinking that `setDBVersion` failing is next to impossible though.